### PR TITLE
Issue 12156 add properties form

### PR DESCRIPTION
**Pull request by woontsoudev**

### CODE REVIEW

File: src/app/portlets/content-types/fields/content-type-fields-drop-zone/content-type-fields-drop-zone.component.ts [hunk 1]
```diff
@@ -41,16 +49,106 @@ export class ContentTypeFieldsDropZoneComponent {
...
+     * @memberof ContentTypeFieldsDropZoneComponent
      */
-    saveFields(): void {
-        this.saveFieldsEvent.emit(this.getFields());
+    saveFields(fieldToSave: Field): void {
+        let fields = this.getFields();
+        // Needs a better implementation
+        fields.map(field => {
+            if (!field.id && !(fieldToSave.clazz === TAB_DIVIDER.clazz || fieldToSave.clazz === LINE_DIVIDER.clazz)) {
+                field = Object.assign(field, fieldToSave);
```
Comment (fmontes) [inline src/app/portlets/content-types/fields/content-type-fields-drop-zone/content-type-fields-drop-zone.component.ts]: You don't need this condition, both cases is doing the same.

File: src/app/portlets/content-types/fields/content-type-fields-drop-zone/content-type-fields-drop-zone.component.ts [hunk 2]
```diff
@@ -41,16 +49,106 @@ export class ContentTypeFieldsDropZoneComponent {
...
+
+    /**
+     * Get the field to be edited
+     * @param {Field} fieldToEdit
+     * @memberof ContentTypeFieldsDropZoneComponent
+     */
+    editField(fieldToEdit: Field): void {
+        let fields = this.getFields();
+        // Needs a better implementation
+        fields.map((field) => {
```
Comment (fmontes) [inline src/app/portlets/content-types/fields/content-type-fields-drop-zone/content-type-fields-drop-zone.component.ts]: You don't need a ```map``` here, ```map``` will return a new array of whatever you return, but in this case you are only looking for a match, I think you can use ```forEach``` instead.

File: src/app/portlets/content-types/fields/content-type-fields-drop-zone/content-type-fields-drop-zone.component.ts [hunk 3]
```diff
@@ -41,16 +49,106 @@ export class ContentTypeFieldsDropZoneComponent {
...
+    }
+
+    /**
+     * Set the field to be edited
+     * @memberof ContentTypeFieldsDropZoneComponent
+     */
+    setDroppedField(): void {
+        let fields = this.getFields();
+        // Needs a better implementation
+        fields.map(field => {
```
Comment (fmontes) [inline src/app/portlets/content-types/fields/content-type-fields-drop-zone/content-type-fields-drop-zone.component.ts]: ```forEach```

File: src/app/portlets/content-types/fields/content-type-fields-drop-zone/content-type-fields-drop-zone.component.ts [hunk 4]
```diff
@@ -41,16 +49,106 @@ export class ContentTypeFieldsDropZoneComponent {
...
+            if (!field.id && !(field.clazz === TAB_DIVIDER.clazz || field.clazz === LINE_DIVIDER.clazz)) {
```
Comment (fmontes) [inline src/app/portlets/content-types/fields/content-type-fields-drop-zone/content-type-fields-drop-zone.component.ts]: We can move this conditional to a method and give it a proper name, like ```isNotRowOrColumn``` or ```isField```.

File: src/app/portlets/content-types/fields/content-type-fields-drop-zone/content-type-fields-drop-zone.component.html [hunk 5]
```diff
@@ -6,13 +6,21 @@
...
     <div *ngIf="!fieldRows.length"
          class="content-type-fields-drop-zone__message">Drop fields here
     </div>
-</div>  
+</div>
+<p-dialog header="Field Properties" [(visible)]="displayDialog" modal="modal" width="300" [responsive]="true">
+    <content-type-fields-properties-form #fieldPropertiesForm (fieldData)="saveFields($event)" [formFieldData]="formData"></content-type-fields-properties-form>
+    <p-footer>
+        <button type="button" pButton (click)="fieldPropertiesForm.saveFieldProperties(fieldPropertiesForm.form.value, fieldPropertiesForm.form.valid)" label="Save"></button>
+        <button type="button" pButton (click)="removeFieldsWithoutId()" label="Cancel"></button>
```
Comment (freddyucv) [inline src/app/portlets/content-types/fields/content-type-fields-drop-zone/content-type-fields-drop-zone.component.html]: Save and Cancel must be I18N messages

File: src/app/portlets/content-types/fields/content-type-fields-drop-zone/content-type-fields-drop-zone.component.ts [hunk 6]
```diff
@@ -41,16 +49,106 @@ export class ContentTypeFieldsDropZoneComponent {
...
+                this.formData = field;
+            }
+        });
     }
 
+    /**
+     * Show the properties field dialog
+     * @memberof ContentTypeFieldsDropZoneComponent
+     */
+    showDialog(): void {
```
Comment (fmontes) [inline src/app/portlets/content-types/fields/content-type-fields-drop-zone/content-type-fields-drop-zone.component.ts]: I think that instead or having2 methods to handle thid, you can create one:

```
toggleDialog() {
    this.displayDialog = !this.displayDialog;
}
```

If not... please use verbs: show/hide.

File: src/app/portlets/content-types/fields/content-type-fields-drop-zone/content-type-fields-drop-zone.component.ts [hunk 7]
```diff
@@ -41,16 +49,106 @@ export class ContentTypeFieldsDropZoneComponent {
...
+    }
+
+    /**
+     * Remove the last dropped field added without ID
+     * @memberof ContentTypeFieldsDropZoneComponent
+     */
+    removeFieldsWithoutId(): void {
+        let fieldRows: any = this.fieldRows;
+
+        fieldRows.forEach((row) => {
```
Comment (fmontes) [inline src/app/portlets/content-types/fields/content-type-fields-drop-zone/content-type-fields-drop-zone.component.ts]: Three ```forEach```? this is satan stuff. There is a better way of doing this with lodash.

File: src/app/portlets/content-types/fields/content-type-fields-drop-zone/content-type-fields-drop-zone.component.ts [hunk 8]
```diff
@@ -62,6 +160,13 @@ export class ContentTypeFieldsDropZoneComponent {
         return result;
     }
 
+    /**
```
Comment (fmontes) [inline src/app/portlets/content-types/fields/content-type-fields-drop-zone/content-type-fields-drop-zone.component.ts]: Private methods doesn't need doc.

File: src/app/portlets/content-types/fields/content-type-fields-drop-zone/content-type-fields-drop-zone.component.ts [hunk 9]
```diff
@@ -41,16 +49,106 @@ export class ContentTypeFieldsDropZoneComponent {
...
+     * @param {Field[]} fields
+     * @returns {Field[][]}
+     * @memberof ContentTypeFieldsDropZoneComponent
+     */
     private splitFieldsByLineDiveder(fields: Field[]): Field[][] {
         let result: Field[][] = [];
         let currentFields: Field[];
-
         fields.forEach(field => {
             if (field.clazz === LINE_DIVIDER.clazz) {
```
Comment (fmontes) [inline src/app/portlets/content-types/fields/content-type-fields-drop-zone/content-type-fields-drop-zone.component.ts]: We should move all this: ```field.clazz === LINE_DIVIDER.clazz``` to methods like:

```
isRow()
isColum()
isField()
```

There are to many everywhere.

File: src/app/portlets/content-types/fields/content-type-fields-properties-form/content-type-fields-properties-form.component.html [hunk 10]
```diff
@@ -0,0 +1,13 @@
+<form [formGroup]="form" (ngSubmit)="saveFieldProperties(form.value, form.valid)" novalidate>
```
Comment (fmontes) [inline src/app/portlets/content-types/fields/content-type-fields-properties-form/content-type-fields-properties-form.component.html]: You don't need to pass 2 params, pass just ```form``` and inside the method get the props you need.

File: src/app/portlets/content-types/fields/content-type-fields-properties-form/content-type-fields-properties-form.component.html [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+    <div class="ui-g">
+        <div class="ui-g-12">
+            <div class="form__group">
+                <label class="form__label">Name</label>
+                <input pInputText type="text" formControlName="name">
+                <small [hidden]="form.controls.name.valid || (form.controls.name.pristine && !submitted)">
```
Comment (fmontes) [inline src/app/portlets/content-types/fields/content-type-fields-properties-form/content-type-fields-properties-form.component.html]: There is a component for this error messages, ```content-type-form.component``` use it.

File: src/app/portlets/content-types/fields/content-type-fields-properties-form/content-type-fields-properties-form.component.html [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+                    Name is required (minimum 5 characters).
```
Comment (fmontes) [inline src/app/portlets/content-types/fields/content-type-fields-properties-form/content-type-fields-properties-form.component.html]: Needs translation.

File: src/app/portlets/content-types/fields/content-type-fields-properties-form/content-type-fields-properties-form.component.ts [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+            this.fieldData.emit(value);
+        }
+    }
+
+    /**
+     * Create the form fields
+     * @private
+     * @memberof ContentTypeFieldsPropertiesFormComponent
+     */
+    private initFormGroup(): void {
```
Comment (fmontes) [inline src/app/portlets/content-types/fields/content-type-fields-properties-form/content-type-fields-properties-form.component.ts]: Private doesn't need doc.

File: src/app/portlets/content-types/fields/content-type-fields-row/content-type-fields-row.component.html [hunk 14]
```diff
@@ -11,11 +11,10 @@
         data-drag-type="target"
         dragula="fields-bag">
        
-        <content-type-field-dragabble-item
-            *ngFor="let field of column.fields"
-            [field]="field"
-            (remove)="removeField($event)">
-
+        <content-type-field-dragabble-item *ngFor="let field of column.fields"
```
Comment (fmontes) [inline src/app/portlets/content-types/fields/content-type-fields-row/content-type-fields-row.component.html]: Leave the indentention as it was and keep that pattern.

File: src/app/portlets/content-types/fields/content-type-fields-row/content-type-fields-row.component.html [hunk 15]
```diff
@@ -11,11 +11,10 @@
...
+                                            [field]="field"
+                                            (remove)="removeField($event)"
+                                            (edit)="fieldToEdit($event)">
```
Comment (fmontes) [inline src/app/portlets/content-types/fields/content-type-fields-row/content-type-fields-row.component.html]: Don't need the method, can do directly:

```(edit)="this.editField.emit(field)"```

File: src/app/portlets/content-types/fields/content-type-fields-properties-form/content-type-fields-properties-form.component.html (hunk 12 above)
Comment (freddyucv) [inline src/app/portlets/content-types/fields/content-type-fields-properties-form/content-type-fields-properties-form.component.html]: no matter, it is going to change a lot

**Comment by fmontes** [comment]
I merged this one, still need more work but @freddyucv need the code to keep moving so he's going to finish it and @woontsoudev will be finishing the tests here.

**Pull request merged**